api-wrapper/query.py

```diff
@@ -0,0 +1,225 @@
+    with open('query.yaml') as f:
+        config = yaml.load(f, yaml.Loader)
+    
+    # config validation
+    if getConfigKey(config, 'query', 'query') == None:
+        raise Exception('Query not defined')
+    if getConfigKey(config, 'query', 'type') == None:
+        raise Exception('Query type not defined')
+    if getConfigKey(config, 'dataset') == None:
+        raise Exception('Dataset not defined')
+
+    limit = getConfigKey(config, 'query', 'limit')
+    if limit == None:
+        limit = 100
+        config['query']['limit'] = limit
+        logger.info('Limit set to {}'.format(limit))
+        print('Limit set to {}'.format(limit))
+    
+    # parse the dates
+    fmt = '%Y-%m-%dT%H:%M:%SZ'
+    st = getConfigKey(config, 'timing', 'start')
+    if st == None:
+        raise Exception('start time not defined')
+    et = getConfigKey(config, 'timing', 'end')
+    if et == None:
+        raise Exception('end time not defined')
+    
+    # Local times
+    startTime = dateparser.parse(st)
+    endTime = dateparser.parse(et)
+    if startTime == None or endTime == None:
+        raise Exception('Failed to parse times')
+    
+    startTime = startTime.astimezone()
+    endTime = endTime.astimezone()
+    now = datetime.now().astimezone()
+    if endTime < startTime or endTime > now:
+        raise Exception('Time travel is not possible')
+    
+    # Convert to UTC
+    config['timing']['start'] = startTime.astimezone(timezone.utc).strftime(fmt)
+    config['timing']['end'] = endTime.astimezone(timezone.utc).strftime(fmt)
+    #print(config)
+
+    queryConfig = config
+    logger.info('Query loaded OK')
+
+def getAPIURL(path):
+    base = getConfigKey(queryConfig, 'endpoint')
+    if base == None:
+        logger.fatal('Grepr endpoint not set in config file')
+        raise Exception('Endpoint no set')
+    
+    if path[0] != '/':
+        path = '/' + path
+    url = urljoin(base, '/api{}'.format(path))
+    logger.info('API URL {}'.format(url))
+    return url
+
+# get all datasets and save the results
+def loadDatasets():
+    global dataSets
+
+    logger.info('Loading datasets')
+    headers = {
+        'Content-Type': 'application/json',
+        'Authorization': 'Bearer {}'.format(authToken['access_token'])
+    }
+
+    url = getAPIURL('/v1/datasets')
+    r = requests.get(url, headers=headers)
+    print('Datasets Status {}'.format(r.status_code))
+    r.raise_for_status()
+
+    dataSets = r.json()
+    logger.info('datasets saved {}'.format(len(dataSets)))
+
+# look up dataset id by name
+def getDatasetID(name):
+    id = None
+    for d in dataSets:
+        if d['name'] == name:
+            id = d['id']
+            break
+    
+    logger.info('Data set {} : {}'.format(name, id))
+    return id
+
+def getConfigKey(data, *keys):
+    if data and keys:
+        element = keys[0]
+        if element:
+            value = data.get(element)
+            return value if len(keys) == 1 else getConfigKey(value, *keys[1:])
```

Comment:
The recursive function is clever but hard to understand. Consider a simpler approach:
```suggestion
def getConfigKey(data, *keys):
    for key in keys:
        if not isinstance(data, dict):
            return None
        data = data.get(key)
        if data is None:
            return None
    return data
```